libraries/persistent_store/src/driver.rs

```diff
@@ -0,0 +1,630 @@
+        &mut self.store
+    }
+
+    /// Provides read-only access to the model.
+    pub fn model(&self) -> &StoreModel {
+        &self.model
+    }
+
+    /// Applies a store operation to the store and model without interruption.
+    pub fn apply(&mut self, operation: StoreOperation) -> Result<(), StoreInvariant> {
+        let (deleted, store_result) = self.store.apply(&operation);
+        let model_result = self.model.apply(operation);
+        if store_result != model_result {
+            return Err(StoreInvariant::DifferentResult {
+                store: store_result,
+                model: model_result,
+            });
+        }
+        self.check_deleted(&deleted)?;
+        Ok(())
+    }
+
+    /// Applies a store operation to the store and model with a possible interruption.
+    pub fn partial_apply(
+        mut self,
+        operation: StoreOperation,
+        interruption: StoreInterruption,
+    ) -> Result<(Option<StoreError>, StoreDriver), (Store<BufferStorage>, StoreInvariant)> {
+        self.store
+            .storage_mut()
+            .arm_interruption(interruption.delay);
+        let (deleted, store_result) = self.store.apply(&operation);
+        Ok(match store_result {
+            Err(StoreError::NoLifetime) => return Err((self.store, StoreInvariant::NoLifetime)),
+            Ok(()) | Err(StoreError::NoCapacity) | Err(StoreError::InvalidArgument) => {
+                self.store.storage_mut().disarm_interruption();
+                let model_result = self.model.apply(operation);
+                if store_result != model_result {
+                    return Err((
+                        self.store,
+                        StoreInvariant::DifferentResult {
+                            store: store_result,
+                            model: model_result,
+                        },
+                    ));
+                }
+                if store_result.is_ok() {
+                    if let Err(invariant) = self.check_deleted(&deleted) {
+                        return Err((self.store, invariant));
+                    }
+                }
+                (store_result.err(), StoreDriver::On(self))
+            }
+            Err(StoreError::StorageError) => {
+                let mut driver = StoreDriverOff {
+                    storage: self.store.into_storage(),
+                    model: self.model,
+                    complete: None,
+                };
+                driver.storage.corrupt_operation(interruption.corrupt);
+                let mut model = driver.model.clone();
+                if model.apply(operation).is_ok() {
+                    driver.complete = Some(Complete { model, deleted });
+                }
+                (None, StoreDriver::Off(driver))
+            }
+            Err(error) => return Err((self.store, StoreInvariant::StoreError(error))),
+        })
+    }
+
+    /// Returns a mapping from delay time to number of modified bits.
+    pub fn delay_map(
```

Comment:
This function (here and above) as a bit of a mystery to me. What is it used for? Why is a `0` added? I feel like it lacks documentation.

Reply:
Added more documentation.